src/CanvasRenderingContext2d.cc

```diff
@@ -1266,41 +1280,74 @@ NAN_METHOD(Context2d::DrawImage) {
   // Start draw
   cairo_save(ctx);
 
-  // Scale src
-  float fx = (float) dw / sw;
-  float fy = (float) dh / sh;
+  cairo_matrix_t matrix;
+  double transforms[6];
+  cairo_get_matrix(context->context(), &matrix);
+  decompose_matrix(matrix, transforms);
+  // extract the scale value from the current transform so that we know how many pixels we
+  // need for our extra canvas in the drawImage operation.
+  double current_scale_x = transforms[1];
+  double current_scale_y = transforms[2];
+  float extra_dx = 0;
+  float extra_dy = 0;
+  float fx = (float) dw / sw * current_scale_x; // transforms[1] is scale on X
+  float fy = (float) dh / sh * current_scale_y; // transforms[2] is scale on X
   bool needScale = dw != sw || dh != sh;
   bool needCut = sw != source_w || sh != source_h || sx < 0 || sy < 0;
-  bool needCairoClip = sx < 0 || sy < 0 || sw > source_w || sh > source_h;
-
+  bool needAdditionalCut = sx < 0 || sy < 0 || sw > source_w || sh > source_h;
```

Comment:
I don't quite understand `needCut` vs. `needAdditionalCut`. Both of them test `sx < 0 || sy < 0`. `sw != source_w` is also true for all cases where `sw > source_w` (ignoring `NaN`, which shouldn't be encountered here).

Reply:
no with the code where it is now, those booleans can become 1. Probably before there was extra code paths to follow in a case or the other.

Reply:
@zbjornson @asturur  it looks to me like `needAdditionalCut` is a renaming of `needCairoClip`, which was introduced to fix #1249 in this related PR: #1251.

I think both this current PR and #1251 are fixes for (among other things) issues that were introduced with the use of `CAIRO_EXTEND_REFLECT`.

Please excuse comment from left-field here :) I hope it's helpful history.

Reply:
Yes i introduced both, they can be unified. Before the code had different srategies depending on what was the reason of the cut.